File: data-skill/templates/r/tte_cloning.R
#!/usr/bin/env Rscript
# Target Trial Emulation - Cloning Template
# Purpose: Emulate multiple treatment strategies using cloning

library(tidyverse)

# ============================================================================
# USER INPUT: Define your trial
# ============================================================================
data_file <- "data.csv"

# Trial components
# 1. Eligibility
eligibility_filter <- "age >= 18"

# 2. Index date (time zero)
index_date_var <- "baseline_date"

# 3. Treatment strategies to compare
# Strategy A: Always treat
# Strategy B: Never treat
# Strategy C: As observed

# 4. Follow-up (days)
follow_up_days <- 365  # 1 year follow-up

# Outcome variable
outcome_var <- "outcome"
# ============================================================================

# Load data
df <- read_csv(data_file)

# ============================================================================
# STEP 1: Define eligibility
# ============================================================================
eligible <- df %>%
  filter(age >= 18) %>%
  filter(!is.na(baseline_covariate))

cat("Eligible population:", nrow(eligible), "subjects\n")

# ============================================================================
# STEP 2: Create cloned dataset
# ============================================================================
# Create multiple copies of each person under different strategies

# Strategy A: Everyone gets treatment
strategy_A <- eligible %>%
  select(id, all_of(index_date_var), treatment, all_of(outcome_var), matches_eligible) %>%
  mutate(
    strategy = "A",
    assigned_treatment = 1,
    assigned_outcome = outcome_var  # Use observed outcome
  )

# Strategy B: No one gets treatment
strategy_B <- eligible %>%
  select(id, all_of(index_date_var), treatment, all_of(outcome_var), matches_eligible) %>%
  mutate(
    strategy = "B",
    assigned_treatment = 0,
    assigned_outcome = outcome_var  # Use observed outcome
  )

# Strategy C: As actually observed
strategy_C <- eligible %>%
  select(id, all_of(index_date_var), treatment, all_of(outcome_var), matches_eligible) %>%
  mutate(
    strategy = "C",
    assigned_treatment = treatment,
    assigned_outcome = outcome_var  # Use observed outcome
  )

# Combine all strategies
tte_data <- bind_rows(strategy_A, strategy_B, strategy_C)

# ============================================================================
# STEP 3: Calculate follow-up time
# ============================================================================
tte_data <- tte_data %>%
  mutate(
    index_date = as.Date(!!index_date_var),
    follow_up_days = as.numeric(as.Date(end_date) - index_date)
  ) %>%
  # Filter to valid follow-up window
  filter(follow_up_days >= 0, follow_up_days <= follow_up_days) %>%
  # Censor at end of follow-up
  mutate(
    event = ifelse(is.na(event), 0, 1),
    censored = ifelse(follow_up_days < follow_up_days, 1, 0)
  )

# ============================================================================
# STEP 4: Analyze outcomes by strategy
# ============================================================================
results <- tte_data %>%
  group_by(strategy) %>%
  summarise(
    n = n(),
    event_count = sum(event),
    person_time = sum(follow_up_days, na.rm = TRUE),
    cumulative_incidence = sum(event) / n(),
    .groups = "drop"
  )

cat("\n========================================\n")
cat("TARGET TRIAL EMULATION RESULTS\n")
cat("========================================\n\n")

print(results)

cat("\nIncidence rates (per 1000 person-years):\n")
results %>%
  mutate(
    rate_per_1000_py = (event_count / person_time) * 1000 * 365
  ) %>%
  select(strategy, rate_per_1000_py) %>%
  print()

cat("\n========================================\n")

# Risk comparison
risk_A <- results$cumulative_incidence[results$strategy == "A"]
risk_B <- results$cumulative_incidence[results$strategy == "B"]

if (!is.na(risk_A) && !is.na(risk_B)) {
  risk_ratio <- risk_A / risk_B
  risk_difference <- risk_A - risk_B

  cat("Risk Ratio (A vs B):", round(risk_ratio, 3), "\n")
  cat("Risk Difference (A vs B):", round(risk_difference, 3), "\n")
}

# ============================================================================
# SAVE RESULTS
# ============================================================================
write_csv(results, "tte_results.csv")
ggsave("tte_comparison.png", width = 8, height = 6, dpi = 300)

File: data-skill/templates/python/tte_cloning.py
#!/usr/bin/env python3
# Target Trial Emulation (TTE) Template
# Purpose: Emulate target trial by cloning dataset for different treatment strategies

import pandas as pd
import numpy as np

# ============================================================================
# USER INPUT: Define your variables
# ============================================================================
# File path to data
data_path = "your_data.csv"

# Eligibility criteria
# Example: age >= 18 and no prior treatment
eligibility_condition = "age >= 18 & prior_treatment == 0"

# Treatment variable (binary: 0 = control, 1 = treated)
treatment_var = "treatment"

# Outcome variable
outcome_var = "outcome"

# Follow-up time variable (optional)
time_var = "followup_time"

# Confounders for adjustment
confounders = ["age", "sex", "baseline_score"]

# ============================================================================
# LOAD AND PREPARE DATA
# ============================================================================
df = pd.read_csv(data_path)

# Convert condition string to pandas query format
# Replace R-style operators with Python equivalents
eligibility_query = eligibility_condition.replace("&", "and").replace("|", "or")

# ============================================================================
# COMPONENT 1: ELIGIBILITY
# ============================================================================
eligible = df.query(eligibility_query).copy()
print(f"Original N: {len(df)}")
print(f"Eligible N: {len(eligible)}")

# ============================================================================
# COMPONENT 2: TREATMENT STRATEGIES
# ============================================================================
# Define strategies to compare
# Strategy A: Everyone gets treatment
# Strategy B: No one gets treatment
# Strategy C: As actually observed

# Strategy A: Everyone gets treatment
strategy_A = eligible.copy()
strategy_A['strategy'] = 'A'
strategy_A['assigned_treatment'] = 1
strategy_A['assigned_outcome'] = strategy_A[outcome_var]  # Observed outcome

# Strategy B: No one gets treatment
strategy_B = eligible.copy()
strategy_B['strategy'] = 'B'
strategy_B['assigned_treatment'] = 0
strategy_B['assigned_outcome'] = strategy_B[outcome_var]  # Observed outcome

# Strategy C: As actually observed
strategy_C = eligible.copy()
strategy_C['strategy'] = 'C'
strategy_C['assigned_treatment'] = strategy_C[treatment_var]
strategy_C['assigned_outcome'] = strategy_C[outcome_var]

# ============================================================================
# COMPONENT 3: TREATMENT ASSIGNMENT
# ============================================================================
# Combine strategies
cloned_df = pd.concat([strategy_A, strategy_B, strategy_C], ignore_index=True)

# In TTE, assignment is deterministic by strategy
# For causal estimation, you would need to apply:
# - IPW (Inverse Probability Weighting)
# - G-formula (Standardization)
# - Doubly robust methods

# Example: Simple g-formula estimator
# Model outcome ~ treatment + confounders for each strategy
from sklearn.linear_model import LinearRegression

# Fit outcome model on observed data (strategy C)
observed = cloned_df[cloned_df['strategy'] == 'C']
X = observed[[treatment_var] + confounders]
y = observed[outcome_var]

model = LinearRegression()
model.fit(X, y)

# Predict potential outcomes for each strategy
for strategy in ['A', 'B']:
    strategy_data = cloned_df[cloned_df['strategy'] == strategy].copy()
    X_strategy = strategy_data[[treatment_var] + confounders]
    cloned_df.loc[cloned_df['strategy'] == strategy, 'predicted_outcome'] = model.predict(X_strategy)

# ============================================================================
# COMPONENT 4: OUTCOMES
# ============================================================================
# Summarize outcomes by strategy
summary = cloned_df.groupby('strategy').agg({
    outcome_var: 'mean',
    'assigned_outcome': 'mean'
}).round(3)

print("\n" + "="*50)
print("OUTCOME COMPARISON BY STRATEGY")
print("="*50 + "\n")
print(summary)

# Calculate treatment effect (comparing A vs B)
outcome_A = cloned_df[cloned_df['strategy'] == 'A']['assigned_outcome'].mean()
outcome_B = cloned_df[cloned_df['strategy'] == 'B']['assigned_outcome'].mean()
effect = outcome_A - outcome_B

print(f"\nCausal Effect (A vs B): {effect:.3f}")

# ============================================================================
# SAVE CLONED DATASET
# ============================================================================
cloned_df.to_csv("tte_cloned_data.csv", index=False)
print("\nCloned dataset saved to: tte_cloned_data.csv")

print("\n" + "="*50)
print("TTE ANALYSIS COMPLETE")
print("="*50)
print("\nNext steps:")
print("1. Apply appropriate causal estimation method (IPW, G-formula, DR)")
print("2. Assess overlap/positivity")
print("3. Conduct sensitivity analysis")
